api/src/main/java/org/openmrs/api/stream/StreamDataService.java

```diff
@@ -62,42 +63,51 @@ public QueueInputStream() {
 		}
 
 		public QueueOutputStream newQueueOutputStream() {
-			return new QueueOutputStream(blockingQueue, timeoutNanos);
+			return new QueueOutputStream(this);
 		}
 
 		@Override
 		public int read() throws IOException {
 			try {
-				if (streamException != null) {
-					// Rethrow exception if writer failed.
-					throw streamException;
-				}
+				checkStreamException();
 				
+				int result;
 				Integer peek = this.blockingQueue.peek();
 				if (Integer.valueOf(-1).equals(peek)) {
-					return -1;
-				}
-
-				Integer value = this.blockingQueue.poll(this.timeoutNanos, TimeUnit.NANOSECONDS);
-				if (value == null) {
-					// Timeout
-					return -1;
-				} else if (value == -1) {
-					// End of stream. Put the end of stream back in the queue for consistency.
-					this.blockingQueue.clear();
-					if (!this.blockingQueue.offer(-1, timeoutNanos, TimeUnit.NANOSECONDS)) {
-						throw new IOException("Failed to write to full queue");
-					}
-					return -1;
+					result = -1;
 				} else {
-					return 255 & value;
+					Integer value = this.blockingQueue.poll(this.timeoutNanos, TimeUnit.NANOSECONDS);
+					if (value == null) {
+						// Timeout
+						result = -1;
+					} else if (value == -1) {
+						// End of stream. Put the end of stream back in the queue for consistency.
+						this.blockingQueue.clear();
+						if (!this.blockingQueue.offer(-1, timeoutNanos, TimeUnit.NANOSECONDS)) {
+							throw new IOException("Failed to write to full queue");
+						}
+						result = -1;
+					} else {
+						result = 255 & value;
+					}
 				}
+
+				checkStreamException();
```

Comment:
Fixing occasionally failing test due to closing broken stream before reading and thus getting result = -1 (end of stream) which hides the exception.